test/test_build_stream.ts

```diff
@@ -0,0 +1,191 @@
+/**
+ * @license
+ * Copyright 2018 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Bluebird from 'bluebird';
+import { Stream, Readable, Writable, Duplex } from 'stream';
+
+import Builder from '../src/index';
+import { sampleDaemonOutputGenerator, sampleDaemonStreamGenerator }
+	from './test-files/sample_daemon_output';
+
+const assert = require('assert').strict;
+
+/**
+ * This test asserts that the createBuildStream() method writes the expected
+ * data to the stream it returns, also prints how it long it took to write the
+ * tar stream to a mocked docker daemon and how long it took to read and parse
+ * the stream/status data from the mocked daemon.  (Note that there are
+ * additional tests in tests.ts that cover other additional functionality such
+ * as calling the success/failure hooks as expected.)
+ *
+ * The test is run only once so obviously does provide accurate time figures at
+ * all, but importantly it allows developers to play with the implementation of
+ * createBuildStream, and then run the test multiple times for the old and new
+ * implementations to compare the performance impact.
+ * 
+ * As is, it takes under 50ms to run in a devenv VirtualBox VM on my laptop.
+ */
+describe('createBuildStream performance', function() {
+	this.timeout(5000);
+
+	it('should be fast', () => {
+		const streamer = sampleDaemonStreamGenerator();
+		const mockBuilder = new MockBuilder();
+		let startTime: number;
+		let buildStream: Duplex;
```

Comment:
```typescript
const buildStream = Builder.prototype.createBuildStream.call(mockBuilder, {});
```
This will remove the need for the `buildStream!` below.

Reply:
Hmm but the assignment would be in a different scope as the usage... Or I didn't get it.

Reply:
@pdcastro you are using it in this scope (line 81) and you need to add a `!` because tsc can't ensure it will be defined then (because it is assigned in a function that runs in parallel).
Define it here and you won't need the `!`.